### ISSUE TITLE
Add hotkeys for Undo/Apply/Cancel buttons

### ISSUE THREAD

**Issue by mk-pmb** [jellysquid3/Sodium#2769]
### Request Description

Could you please add a way to reliably activate the Undo/Apply/Cancel buttons via keyboard? Maybe like those underlined characters on buttons when you press the Alt key in most traditional desktop UIs. I tried Alt+U, Alt+A, Alt+C to see if maybe it's secretly implemented, but it had no effect.

I'm trying to automatically adjust some settings via simulated keystrokes. One of my problems is navigating to the undo/apply/cancel buttons after I managed to change the settings, especially because the number of focus-enabled (implies: visible) buttons changes based on whether the new settings actually differ from the old ones. My fallback strategy would be to analyze screenshots and simulate mouse input, but I hope this is a very small change for you and could save a lot of code on my side. It may also have minor usability benefits for other users.

**Comment by jellysquid3** [comment]
I'm fine with the idea of underlining a character on the buttons when ALT is held, and allowing the user to jump to that in the focus tree when ALT+LETTER is pressed. This is typical of Windows applications, even if Minecraft doesn't provide it.

**Comment by mk-pmb** [comment]
> allowing the user to jump to that in the focus tree

Thanks! Even better would be if the Alt key combination would not just jump to the button, but actually fire it, like in Windows or Gnome dialogs. That way, I could reliably fire Apply (if available) and then Done. Currently, the different behavior depending on whether I have to press one or two buttons to save and go back to the Options menu, requires [a workaround activating and cancelling ingame chat.](https://github.com/mk-pmb/minecraft-nether-roof-biomes-marker/blob/861d8e282e8c4ed2f1e9c9fbb522edfd9e3d1412/util/save_or_discard_sodium_settings.sh#L34)

> even if Minecraft doesn't provide it.

Indeed it would be very helpful In the Minecraft menus, but there it's less of a problem, because its menu buttons can reliably be focussed with unchanging arrow key sequences.

**Comment by marco-fernandes2006-nofc** [comment]
Hello, I've been following this issue and the discussion around it.

From what I understand, the goal is to improve interaction with the Undo, Apply, and Cancel buttons, and there was some mention of Alt-based navigation or focus highlighting depending on the system.

I’d be interested in working on this, as I’d like to contribute more to the project.

Before starting, I wanted to check if this is still something that should be implemented, and whether anyone is already working on it.

My initial idea would be to focus on adding direct keyboard shortcuts (e.g. Alt+U / Alt+A / Alt+C or similar), as a more straightforward and consistent solution across systems. These shortcuts also seem intuitive and align with existing conventions.

Let me know if this approach makes sense.

**Comment by douira** [comment]
i was not previously tracking this issue in the context of the config ui overhaul but I think it is a good idea.

Another thing, which may be simpler, is to make the escape key activate "undo" when there is an unapplied change, and then pressing escape again once there is no unapplied change should exist the video settings. We have a list of other related features/bugs here: https://github.com/CaffeineMC/sodium/issues/3385

Feel free to comment to claim features you're implementing or if you have any further questions. Thanks for contributing!

**Comment by mk-pmb** [comment]
> and then pressing escape again once there is no unapplied change should exist the video settings.

Please don't mix navigation and actions on the same key. It's annoying for users with sluggish machines, or if any other bug (potentially from another addon) causes the game to freeze for a moment. They may think their first keystroke had been eaten.
Also would this mean you can lose unsaved changes by pressing escape? That would be an additional problem. Escape key is a habitual panic action for users, so whatever it does should always be safe. (We do not have control over what could have caused panic in that situation. We may even have gained input focus by accident.)
For my bot scenario it would add another UI state I'd need to track (what Escape will do), would it?

**Comment by douira** [comment]
Automatically applying all pending changes when you press escape is unintuitive since typically escape implies some sort of "abort" or "leave" action. At the same time, we need to define what happens when the escape key is pressed since currently it simply does nothing, which is also bad because the user expects the screen to close when escape is pressed. Discarding unapplied changes when the escape key is pressed is _also_ unintuitive since in Vanilla pressing escape doesn't revert changes that have been made because the vanilla video settings screen doesn't have the concept of pending changes to begin this. This leaves us with a conundrum that I think the two-press solution nicely solves. I've had an extended conversation about the ergonomics of this with users who were complaining about the escape key currently doing nothing, and they'd rather prefer the escape key applying the settings and closing the screen at once since they would like to make and apply changes without needing to move the mouse to the apply button.

I've never heard anyone report that the UI freezes on user actions. This is not a concern I am aware of or that makes sense in the context of how the screen works.

The ergonomics for typical users are what matters to our decision making here. For automation purposes, I would suggest calling the methods of our config system directly, or something like that.

**Comment by marco-fernandes2006-nofc** [comment]
> Automatically applying all pending changes when you press escape is unintuitive since typically escape implies some sort of "abort" or "leave" action.

I agree that the Esc key shouldn’t automatically apply changes. At the same time, I’m not sure using it to undo changes would be the best fit either.

Because of that, I’m not sure the Esc key should be directly tied to any of the buttons. Instead, it could behave as a "close" action for the menu.

One possible approach would be:

* If there are no pending changes, Esc behaves as it currently does.
* If there are pending changes, it prompts the user to choose whether to apply or discard them before leaving the Video Settings screen.

From some informal feedback I gathered, this approach (using Alt-based shortcuts together with Esc to exit and prompt when there are pending changes) also seems more intuitive and consistent with expected UI behavior.

Going back to the original idea, I think this could work well alongside the Alt-based shortcuts, or even independently if only the shortcuts are implemented.

I also wanted to ask how assignment is handled for issues like this. I’d like to avoid any overlap in case someone else starts working on it.

**Comment by davepusey** [comment]
It's been UI/UX convention for several decades now that Escape = Cancel, so the correct thing it should do is to discard any changes that have not yet been applied.

**Comment by douira** [comment]
I'm not against a basic prompt if you press escape with pending changes and alt-based shortcuts for direct interaction with the buttons. We could even put a checkbox in the prompt with something like "don't show again" that makes it so the escape key always discards changes without asking. As for making the alt-key based navigation more obvious to the user, when the alt key is held down, our UI "framework" should somehow highlight buttons that have keybinds with a line under the letter they refer to. (Referring to what Jellysquid mentioned about how Windows (and many other OSes) work)

**Comment by douira** [comment]
I don't think there's currently anyone else working on this. I recommend you comment on [the tracking issue](https://github.com/CaffeineMC/sodium/issues/3385) with what you plan on doing. It's probably also advisable to correspond with us in the #dev-sodium channel on [our discord server](https://caffeinemc.net/discord).
